planner/core/logical_plan_builder.go

```diff
@@ -1138,96 +1138,193 @@ func (b *PlanBuilder) buildProjection4Union(ctx context.Context, u *LogicalUnion
+		}
+		if rightPlan.Schema().Len() != columnNums {
+			return nil, nil, ErrWrongNumberOfColumnsInSelect.GenWithStackByArgs()
+		}
+		joinPlan := LogicalJoin{JoinType: SemiJoin}.Init(b.ctx, b.getSelectOffset())
+		joinPlan.SetChildren(leftPlan, rightPlan)
+		joinPlan.SetSchema(leftPlan.Schema())
+		joinPlan.names = make([]*types.FieldName, leftPlan.Schema().Len())
+		copy(joinPlan.names, leftPlan.OutputNames())
+		for j := 0; j < len(rightPlan.Schema().Columns); j++ {
+			leftCol, rightCol := leftPlan.Schema().Columns[j], rightPlan.Schema().Columns[j]
+			eqCond, err := expression.NewFunction(b.ctx, ast.NullEQ, types.NewFieldType(mysql.TypeTiny), leftCol, rightCol)
+			if err != nil {
+				return nil, nil, err
+			}
+			joinPlan.EqualConditions = append(joinPlan.EqualConditions, eqCond.(*expression.ScalarFunction))
+		}
+		leftPlan, err = b.buildDistinct(joinPlan, leftPlan.Schema().Len())
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return leftPlan, selects[0].AfterSetOperator, nil
 }
 
-// divideUnionSelectPlans resolves union's select stmts to logical plans.
-// and divide result plans into "union-distinct" and "union-all" parts.
-// divide rule ref: https://dev.mysql.com/doc/refman/5.7/en/union.html
-// "Mixed UNION types are treated such that a DISTINCT union overrides any ALL union to its left."
-func (b *PlanBuilder) divideUnionSelectPlans(ctx context.Context, selects []*ast.SelectStmt) (distinctSelects []LogicalPlan, allSelects []LogicalPlan, err error) {
-	firstUnionAllIdx, columnNums := 0, -1
-	// The last slot is reserved for appending distinct union outside this function.
-	children := make([]LogicalPlan, len(selects), len(selects)+1)
-	for i := len(selects) - 1; i >= 0; i-- {
-		stmt := selects[i]
-		if firstUnionAllIdx == 0 && stmt.IsAfterUnionDistinct {
-			firstUnionAllIdx = i + 1
+func (b *PlanBuilder) buildExcept(ctx context.Context, selects []LogicalPlan, afterSetOpts []*ast.SetOprType) (LogicalPlan, error) {
+	unionPlans := []LogicalPlan{selects[0]}
+	tmpAfterSetOpts := []*ast.SetOprType{nil}
+	columnNums := selects[0].Schema().Len()
+	for i := 1; i < len(selects); i++ {
+		rightPlan := selects[i]
+		if rightPlan.Schema().Len() != columnNums {
+			return nil, ErrWrongNumberOfColumnsInSelect.GenWithStackByArgs()
 		}
+		if *afterSetOpts[i] == ast.Except {
+			leftPlan, err := b.buildUnion(ctx, unionPlans, tmpAfterSetOpts)
+			if err != nil {
+				return nil, err
+			}
+			joinPlan := LogicalJoin{JoinType: AntiSemiJoin}.Init(b.ctx, b.getSelectOffset())
+			joinPlan.SetChildren(leftPlan, rightPlan)
+			joinPlan.SetSchema(leftPlan.Schema())
+			joinPlan.names = make([]*types.FieldName, leftPlan.Schema().Len())
+			copy(joinPlan.names, leftPlan.OutputNames())
+			for j := 0; j < len(rightPlan.Schema().Columns); j++ {
+				leftCol, rightCol := leftPlan.Schema().Columns[j], rightPlan.Schema().Columns[j]
+				eqCond, err := expression.NewFunction(b.ctx, ast.NullEQ, types.NewFieldType(mysql.TypeTiny), leftCol, rightCol)
```

Comment:
Why do we use `NewFieldType(mysql.TypeTiny)` here instead of `leftCol.RetType()`?

Reply:
Because it's an `EQ` function.